test/beast/core/buffers_adaptor.cpp

```diff
@@ -85,6 +113,69 @@ class buffers_adaptor_test : public unit_test::suite
         read_size(ba, 1024);
     }
 
+    template<bool isMutable>
+    void
+    testSubrange()
+    {
+        auto exemplar = std::string("the quick brown fox jumps over the lazy dog");
+
+        auto iterate_test = [&](
+            std::size_t a,
+            std::size_t b,
+            std::size_t c)
+        {
+            static const auto func = "iterate_test";
+            auto buffers = std::vector<net::mutable_buffer>();
+            if (a)
+                buffers.push_back(net::buffer(&exemplar[0], a));
+            if (b - a)
+                buffers.push_back(net::buffer(&exemplar[a], (b - a)));
+            if (c - b)
+                buffers.push_back(net::buffer(&exemplar[b], (c - b)));
+            auto adapter = buffers_adaptor<std::vector<net::mutable_buffer>>(buffers);
+
+
+            using value_type =
+            typename std::conditional<
+                isMutable,
+                net::mutable_buffer,
+                net::const_buffer>::type;
+
+            using maybe_mutable =
+            typename std::conditional<
+                isMutable,
+                buffers_adaptor<std::vector<net::mutable_buffer>>&,
+                buffers_adaptor<std::vector<net::mutable_buffer>> const&>::type;
+
+            auto sub = buffers_adaptor_test_hook::make_subrange(static_cast<maybe_mutable>(adapter));
+            BEAST_EXPECTS(typeid(typename decltype(sub)::value_type) == typeid(value_type), func);
+
+            auto as_string = [](decltype(sub) const &sr) {
+                std::string result;
+                auto first = net::buffer_sequence_begin(sr);
+                auto last = net::buffer_sequence_end(sr);
+                for (; first != last; ++first)
+                {
+                    auto buf = *first;
+                    auto f = reinterpret_cast<const char *>(buf.data());
+                    auto l = f + buf.size();
+                    result.insert(std::end(result), f, l);
```

Comment:
Why `std::end`? The type of `result` is known, this is not a generic context. Nothing is gained by calling `std::end` instead of just using `result.end`. Or even calling `result.append(f, l)`. I would like to understand why you chose one over the other, lets discuss.

Reply:
I use std::end by force of habit, since it works in all contexts so I don't have to waste time reasoning about the underlying container. It's a zero cost abstraction, and I prefer free functions over member functions.
I could have used append, but it's specific to std::string and not generic across all containers, so to me it smells bad because it makes the code logic brittle (if I choose a different container type, I have to re-write logic, which makes no sense).